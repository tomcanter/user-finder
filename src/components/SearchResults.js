// @flow

import React, {
  PropTypes,
} from 'react';
import curry from 'lodash/fp/curry';
import map from 'lodash/fp/map';
import Result from './Result';

type Props = {
  entities: Object,
  query: string,
  ids: Array<number>,
  total: number,
};

const renderSearchResult = curry((entities, item) => {
  const {
    id,
    avatar_url: avatarUrl,
    login,
  } = entities[item];
  return (
    <li key={id}>
      <Result
        username={login}
        avatarUrl={avatarUrl}
      />
    </li>
  );
});

function renderResultsMessage(query: string, total: number) {
  if (total === 0) {return null;}
  return (
    <p>Searching {query} found {total} {total === 1 ? 'result' : 'results'}</p>
  );
}

function SearchResults({entities, ids, query, total}: Props) {
  return (
    <div>
      {renderResultsMessage(query, total)}
      <ul>
        {map(renderSearchResult(entities), ids)}
      </ul>
    </div>
  );
}

SearchResults.propTypes = {
  entities: PropTypes.object.isRequired,
  query: PropTypes.string.isRequired,
  ids: PropTypes.array.isRequired,
  total: PropTypes.number.isRequired,
};

export default SearchResults;
