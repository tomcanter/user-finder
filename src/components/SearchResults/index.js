// @flow

import React, {
  PropTypes,
} from 'react';
import curry from 'lodash/fp/curry';
import map from 'lodash/fp/map';
import {
  StyleSheet,
  css,
} from 'aphrodite/no-important';
import Result from 'components/Result';

type Props = {
  entities: Object,
  searchTerm: string,
  ids: Array<number>,
  total: number,
};

const renderSearchResult = curry((entities, item) => {
  const {
    id,
    avatar_url: avatarUrl,
    login,
  } = entities[item];
  return (
    <li key={id} className={css(styles.SearchResults_item)}>
      <Result
        username={login}
        avatarUrl={avatarUrl}
      />
    </li>
  );
});

function renderResultsMessage(searchTerm: string, total: number, resultLength: number) {
  if (!searchTerm || !resultLength) {return null;}
  return (
    <p><b>{searchTerm}</b> found {total} {total === 1 ? 'result' : 'results'}</p>
  );
}

function SearchResults({entities, ids, searchTerm, total}: Props) {
  return (
    <div className={css(styles.SearchResults)}>
      {renderResultsMessage(searchTerm, total, ids.length)}
      <ul className={css(styles.SearchResults_list)}>
        {map(renderSearchResult(entities), ids)}
      </ul>
    </div>
  );
}

SearchResults.propTypes = {
  entities: PropTypes.object,
  searchTerm: PropTypes.string,
  ids: PropTypes.array,
  total: PropTypes.number,
};

const BREAKPOINT = '@media (min-width: 600px)';

const styles = StyleSheet.create({
  SearchResults: {
    maxWidth: 1280,
    margin: '0 auto',
  },

  SearchResults_list: {
    display: 'flex',
    flexWrap: 'wrap',
  },

  SearchResults_item: {
    flex: '1 0 auto',

    [BREAKPOINT]: {
      ':only-child': {
        flex: 0,
      },
    },
  },
});

export default SearchResults;
