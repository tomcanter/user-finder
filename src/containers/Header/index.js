// @flow

import React, {
  Component,
  PropTypes,
} from 'react';
import {
  StyleSheet,
} from 'aphrodite/no-important';
import Header from 'components/Header';
import Container from 'components/Container';
import connect from './connect';

type Props = {
  onSubmit: Function,
  searchQuery: string,
  searchTerm: string,
  searchForUser: Function,
};

export class HeaderContainer extends Component {

  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    searchQuery: PropTypes.string.isRequired,
    searchTerm: PropTypes.string.isRequired,
    searchForUser: PropTypes.func.isRequired,
  };

  static defaultProps = {
    searchTerm: '',
  };

  constructor(props: Props) {
    super(props);
    this.handleSearchUser(props.searchQuery);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.searchQuery !== this.props.searchQuery) {
      this.handleSearchUser(nextProps.searchQuery);
    }
  }

  handleSearchUser(searchQuery: string): void {
    if (!searchQuery) {return;}
    this.props.searchForUser(searchQuery);
  }

  render() {
    const {
      onSubmit,
      searchTerm,
    } = this.props;

    return (
      <Container rootStyle={styles.HeaderContainer}>
        <Header
          onSubmit={onSubmit}
          searchTerm={searchTerm}
        />
      </Container>
    );
  }

}

const styles = StyleSheet.create({
  HeaderContainer: {
    backgroundColor: '#f4f4f4',
    borderBottom: '1px solid #aaa',
    boxShadow: '0px 0px 2px 1px rgba(0, 0, 0, .2)',
  },
});

export default connect(HeaderContainer);
