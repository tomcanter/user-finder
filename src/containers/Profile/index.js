import React, {
  Component,
} from 'react';
import get from 'lodash/fp/get';
import {
  css,
  StyleSheet,
} from 'aphrodite/no-important';
import Container from 'components/Container';
import ProfileHeader from 'components/ProfileHeader';
import ItemList from 'components/ItemList';
import User from 'components/User';
import Repo from 'components/Repo';
import Loading from 'components/Loading';
import ErrorBoundary from 'components/ErrorBoundary';
import {viewport} from 'theme';
import 'suitcss-utils-flex/lib/flex-sm.css';
import 'suitcss-utils-size/lib/size-sm.css';
import 'suitcss-components-grid';
import connect from './connect';

type Props = {
  followerIds: Array<number>,
  followerIsPending: boolean,
  getProfile: Function,
  userIsPending: boolean,
  userProfile: Object,
  repoEntities: Object,
  repoIds: Array<number>,
  repoIsPending: boolean,
  userEntities: Object,
};

const getUsername = get('match.params.username');

export class ProfileContainer extends Component {

  props: Props;

  constructor(props: Props) {
    super(props);
    this.props.getProfile(getUsername(props));
  }

  componentWillReceiveProps(nextProps: Props) {
    const username = getUsername(this.props);
    const nextUsername = getUsername(nextProps);

    if (username !== nextUsername) {
      this.props.getProfile(nextUsername);
    }
  }

  renderProfile() {
    if (this.props.userIsPending) {
      return (
        <div className={css(styles.Profile_wrapLoading)}>
          <Loading />
        </div>
      );
    }

    const {
      repoIsPending,
      repoEntities,
      repoIds,
      followerIds,
      followerIsPending,
      userEntities,
    } = this.props;

    return (
      <div className={css(styles.Profile_container)}>
        <ErrorBoundary>
          <ProfileHeader {...this.props.userProfile} />
        </ErrorBoundary>
        <div className={`${css(styles.Profile_content)} Grid Grid--withGutter`}>
          <div className={`${css(styles.Profile_repos)} Grid-cell u-sm-size1of2`}>
            <h2 className={css(styles.Profile_contentTitle)}>Repositories</h2>
            <ErrorBoundary>
              <ItemList
                entities={repoEntities}
                ids={repoIds}
                isPending={repoIsPending}
                component={Repo}
              />
            </ErrorBoundary>
          </div>
          <div className="Grid-cell u-sm-size1of2">
            <h2 className={css(styles.Profile_contentTitle)}>Followers</h2>
            <ErrorBoundary>
              <ItemList
                entities={userEntities}
                ids={followerIds}
                isPending={followerIsPending}
                component={User}
              />
            </ErrorBoundary>
          </div>
        </div>
      </div>
    );
  }

  render() {
    return (
      <Container>
        {this.renderProfile()}
      </Container>
    );
  }

}

const styles = StyleSheet.create({
  Profile_wrapLoading: {
    marginTop: 25,
  },

  Profile_repos: {
    marginBottom: 20,

    [viewport.SM]: {
      marginBottom: 0,
    },
  },

  Profile_contentTitle: {
    fontSize: 22,
    fontWeight: 400,
    marginBottom: 25,
  },

  Profile_content: {
    paddingTop: 20,
  },

  Profile_container: {
    paddingTop: 15,
  },
});

export default connect(ProfileContainer);
