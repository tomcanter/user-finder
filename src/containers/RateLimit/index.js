import React from 'react';
import {
  css,
  StyleSheet,
} from 'aphrodite/no-important';
import 'suitcss-utils-flex/lib/flex.css';
import RateLimit from 'components/RateLimit';
import connect from './connect';

type Props = {
  coreLimit: ?number,
  coreRemaining: ?number,
  coreReset: ?string,
  searchLimit: ?number,
  searchRemaining: ?number,
  searchReset: ?string,
};

export function RateLimitContainer(props: Props) {
  const {
    coreLimit,
    coreRemaining,
    coreReset,
    searchLimit,
    searchRemaining,
    searchReset,
  } = props;

  return (
    <div className="u-flex u-flexAlignItemsCenter">
      <p className={css(styles.RateLimitContainer_label)}>API requests</p>
      <div
        className={`${css(styles.RateLimitContainer)} u-flex`}
      >
        <div className={css(styles.RateLimitContainer_item)}>
          <RateLimit
            label="Profile"
            remaining={coreRemaining}
            limit={coreLimit}
            reset={coreReset}
          />
        </div>
        <div className={css(styles.RateLimitContainer_item)}>
          <RateLimit
            label="Search"
            remaining={searchRemaining}
            limit={searchLimit}
            reset={searchReset}
          />
        </div>
      </div>
    </div>
  );
}

const styles = StyleSheet.create({
  RateLimitContainer: {
    backgroundColor: '#333',
    color: '#eee',
    padding: '5px 8px',
    borderRadius: 5,
  },

  RateLimitContainer_label: {
    fontSize: 14,
    marginRight: 10,
  },

  RateLimitContainer_item: {
    ':first-child': {
      marginRight: 5,
    },

    ':last-child': {
      marginLeft: 5,
    },
  },
});

export default connect(RateLimitContainer);
