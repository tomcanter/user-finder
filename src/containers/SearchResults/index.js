// @flow

import React from 'react';
import {
  StyleSheet,
} from 'aphrodite/no-important';
import SearchResults from 'components/SearchResults';
import Pagination from 'components/Pagination';
import ResultsMessage from 'components/ResultsMessage';
import Loading from 'components/Loading';
import Container from 'components/Container';
import connect from './connect';

type Props = {
  isPending: boolean,
  pagination: Object,
  searchTerm: string,
  totalResults: number,
  userEntities: Object,
  userIds: Array<number>,
};

export function SearchResultsContainer(props: Props) {
  const {
    isPending,
    pagination,
    searchTerm,
    totalResults,
    userEntities,
    userIds,
  } = props;

  if (!searchTerm) {return null;}

  if (isPending) {
    return <Loading />;
  }

  return (
    <div>
      <Container
        rootStyle={styles.SearchResultsContainer_message}
      >
        <ResultsMessage
          searchTerm={searchTerm}
          resultsTotal={totalResults}
          pageTotal={userIds.length}
        />
      </Container>
      <Container
        noGutter={true}
        rootStyle={styles.SearchResultsContainer_results}
      >
        <SearchResults
          searchTerm={searchTerm}
          ids={userIds}
          entities={userEntities}
          total={totalResults}
        />
      </Container>
      <Container>
        <Pagination {...pagination} />
      </Container>
    </div>
  );
}

SearchResultsContainer.defaultProps = {
  pagination: {},
  searchTerm: '',
};

const styles = StyleSheet.create({
  SearchResultsContainer_results: {
    marginTop: 15,
  },

  SearchResultsContainer_message: {
    marginTop: 15,
  },
});

export default connect(SearchResultsContainer);
