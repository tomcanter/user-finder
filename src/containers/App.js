// @flow

import React from 'react';
import {
  BrowserRouter as Router,
  Route,
} from 'react-router-dom';

import 'suitcss-base';
import SearchScreen from '../screens/SearchScreen';
import Profile from '../screens/Profile';

export default function App() {
  return (
    <Router>
      <div className="App">
        <Route exact={true} path="/" component={SearchScreen} />
        <Route exact={true} path="/:username" component={Profile} />
      </div>
    </Router>
  );
}
