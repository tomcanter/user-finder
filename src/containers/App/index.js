// @flow

import React from 'react';
import {
  HashRouter as Router,
  Redirect,
  Switch,
} from 'react-router-dom';
import SearchScreen from 'screens/Search';
import Profile from 'screens/Profile';
import DefaultLayout from 'screens/Default';

import 'suitcss-base';
import 'css/global.css';

export default function App() {
  return (
    <Router>
      <Switch>
        <DefaultLayout exact={true} path="/search" component={SearchScreen} />
        <DefaultLayout exact={true} path="/:username" component={Profile} />
        <Redirect to="/search" />
      </Switch>
    </Router>
  );
}
