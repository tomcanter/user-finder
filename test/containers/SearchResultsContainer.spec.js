import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import {StyleSheetTestUtils} from 'aphrodite';

import {SearchResultsContainer} from 'containers/SearchResults';

describe('Component: SearchResultsContainer', () => {

  beforeAll(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterAll(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  describe.skip('when the search prop is not present', () => {
    it('should not call the search action', () => {
      const spy = jest.fn();
      shallow(
        <SearchResultsContainer
          searchTerm={''}
          userEntities={{}}
          userIds={[]}
          totalResults={0}
        />
      );
      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe.skip('when the search prop is present', () => {
    it('should call the search action', () => {
      const spy = jest.fn();
      shallow(
        <SearchResultsContainer
          searchQuery={'?q=test'}
          searchUser={spy}
          onSubmit={jest.fn()}
          userEntities={{}}
          userIds={[]}
          totalResults={0}
        />
      );
      expect(spy.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe.skip('when the search prop changes', () => {
    it('should call the search action with the new value', () => {
      const spy = jest.fn();
      const wrapper = shallow(
        <SearchResultsContainer
          searchQuery={'?q=test'}
          searchUser={spy}
          onSubmit={jest.fn()}
          userEntities={{}}
          userIds={[]}
          totalResults={0}
        />
      );
      expect(spy.mock.calls[0]).toMatchSnapshot();
      wrapper.setProps({search: '?q=foobar'});
      expect(spy.mock.calls[1]).toMatchSnapshot();
    });
  });

  describe('when the searchTerm is falsey', () => {
    it('should not render the pagination or results', () => {
      const component = renderer.create(
        <SearchResultsContainer
          searchTerm={''}
          userEntities={{}}
          userIds={[]}
          totalResults={0}
        />
      );
      expect(component).toMatchSnapshot();
    });
  });

});
