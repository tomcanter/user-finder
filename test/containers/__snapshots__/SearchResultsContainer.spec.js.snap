exports[`Component: SearchResultsContainer when the searchTerm is falsey should not render the pagination or results 1`] = `null`;
